### DISCUSSION TITLE
MIDI thru?

### DISCUSSION THREAD

**Discussion by egasimus** [bangcorrupt/hacktribe#189] [Ideas]
So in the dark of the night I made myself a second TRS-MIDI cable to connect Keystep->Electribe->OtherSynth and only then did I realize - the E2 doesn't actually do MIDI thru... does it? Bummer.

I wonder if that's on the roadmap? I also wonder if I could hack it in myself to learn proper Ghidra usage... Seems simple enough, just copy inputs to outputs, but of course with embedded there are always surprises...

**Comment by bangcorrupt** [comment]
MIDI Thru is planned.  I think it could be relatively easy to implement in the CPU, I have most of the CPU MIDI IO figured.  ~I want to try implementing Thru in the MCU, before the MIDI data is passed to the CPU.~ (MIDI does not pass via the MCU).

2024-12-31: MIDI Thru is in the `main` branch.

**Reply by egasimus** [reply]
Fantastic! :rocket:

**Reply by bangcorrupt** [reply]
I realised this is switched off by default.  There is a hidden global setting, accessible via MIDI NRPN.

```
NRPN-MSB = 0x03 # Global settings
NRPN-LSB = 0x00 # Byte index MSB
DATA-MSB = 0x2c # Byte index LSB
DATA-LSB = 0x01 # Enabled
```

The full message to enable MIDI Thru:

```
0xb0, 0x63, 0x03, 0xb0, 0x62, 0x00, 0xb0, 0x06, 0x2c, 0xb0, 0x26, 0x01
```

Once enabled, go into the global settings menu and press `Write` to make the change persistent.

**Reply by richardarnezjames** [reply]
Christmas on NYE, truly a
hacktribe miracle.
